File: java-proxy/src/main/java/com/proxy/ResponseHandler.java
package com.proxy;

import io.undertow.server.HttpServerExchange;
import io.undertow.util.Headers;

import java.nio.charset.StandardCharsets;

public class ResponseHandler {

    public static boolean hasCustomHandler(String path) {
        switch (path) {
            case "/wheredidicomefrom":
                return true;
            default:
                return false;
        }
    }

    public static void handle(String path, byte[] originalBody, HttpServerExchange exchange) {
        switch (path) {
            case "/wheredidicomefrom":
                handleWheredidicomefrom(originalBody, exchange);
                break;
            default:
                exchange.getResponseSender().send(new String(originalBody, StandardCharsets.UTF_8));
                break;
        }
    }

    private static void handleWheredidicomefrom(byte[] originalBody, HttpServerExchange exchange) {
        String originalHtml = new String(originalBody, StandardCharsets.UTF_8);
        String injectedHtml = "<p style=\"color: red; font-weight: bold;\">Injected by the Java Proxy!</p>";
        String modifiedHtml = originalHtml.replace("</body>", injectedHtml + "</body>");

        exchange.getResponseHeaders().put(Headers.CONTENT_LENGTH, (long) modifiedHtml.getBytes(StandardCharsets.UTF_8).length);
        exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/html");

        Logger.log(Logger.Level.INFO, "Successfully injected custom HTML into response");

        exchange.getResponseSender().send(modifiedHtml);
    }
}

File: node-proxy/src/response-handler.ts
import { IncomingMessage, ServerResponse } from 'http';
import { logger, LogLevel } from './logger';

type ResponseHandler = (proxyRes: IncomingMessage, req: IncomingMessage, res: ServerResponse) => void;

export const customHandlers: { [key: string]: ResponseHandler } = {
  '/wheredidicomefrom': handleWheredidicomefrom,
};

export function handleCustomResponse(proxyRes: IncomingMessage, req: IncomingMessage, res: ServerResponse) {
  const handler = customHandlers[req.url || ''];
  if (handler) {
    handler(proxyRes, req, res);
  }
}

function handleWheredidicomefrom(proxyRes: IncomingMessage, req: IncomingMessage, res: ServerResponse) {
  const body: Buffer[] = [];
  proxyRes.on('data', chunk => {
    body.push(chunk);
  });
  proxyRes.on('end', () => {
    const originalBody = Buffer.concat(body).toString();
    const injectedHtml = '<p style="color: blue; font-weight: bold;">Injected by the Node.js Proxy!</p>';
    const modifiedBody = originalBody.replace('</body>', `${injectedHtml}</body>`);

    res.writeHead(200, {
      'Content-Type': 'text/html',
      'Content-Length': Buffer.byteLength(modifiedBody),
    });
    res.end(modifiedBody);
    logger(LogLevel.INFO, 'Successfully injected custom HTML into response');
  });
}
